node/workflow-api.md

```diff
@@ -0,0 +1,322 @@
+# 3rd revision of the Typescript SDK Workflow API
+
+### Current API
+
+```ts
+import { Trigger } from '@temporalio/workflow';
+
+export type MyWorkflow = (
+  arg1: number,
+  arg2: string
+) => {
+  execute(): Promise<void>;
+  signals: {
+    unblock(): void;
+  };
+  queries: {
+    isBlocked(): boolean;
+  };
+};
+
+export const myWorkflow: MyWorkflow = (arg1, arg2) => {
+  const unblocked = new Trigger<void>();
+  let isBlocked = true;
+  return {
+    // Main workflow logic
+    async execute() {
+      await unblocked;
+      isBlocked = false;
+    },
+    signals: {
+      unblock() {
+        unblocked.resolve();
+      },
+    },
+    queries: {
+      isBlocked() {
+        return isBlocked;
+      },
+    },
+  };
+};
+```
+
+#### Issues
+
+- Types must be explicitly declared (To make [eslint happy](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/explicit-module-boundary-types.md))
+- Very verbose
+- Trigger (ResolvablePromise) is a foreign concept to JS developers
+- Trigger doesn't provide all of the functionality for this workflow, an additional `isBlocked` boolean variable is required
+- Signals and queries must be defined at the top level in order for types to be inferred.
+
+#### Advantages
+
+- Types for all Workflow handlers can be inferred at Workflow handle creation
+
+### Goals for the redesign
+
+- Maintain type inference capabilities - Mandatory for the Typescript SDK
+- Dynamic signals and queries registration first - A single API for Workflow handler registration
+- Componentization of related functionality - increase reusability
+- Implicit interface declarations - less boilerplate
+- No global variables - to support running Workflows in less isolated runtimes
+
+### New proposal
+
+#### Remove the `Trigger` API in favor of `condition`
+
+`condition` is similar to `Workflow.await` in Java, it returns a Promise and accepts a closure which resolve the Promise when it returns `true`.
+
+```ts
+import { condition } from '@temporalio/workflow';
+
+let isBlocked = true;
+await condition(() => !isBlocked);
+```
+
+#### Remove the Workflow factory wrapper and always use dynamic handler registration
+
+`workflows.ts`
+
+```ts
+import { Signal, Query, condition } from '@temporalio/workflow';
+
+// Define the signal / query with the type signature at the top level to share the definition for handle creation
+export const unblockSignal = new Signal<[] /* args */>('unblock');
+export const isBlockedQuery = new Query<[] /* args */, boolean /* return type */>('isBlocked');
```

Comment:
If querying a mutable value is a common enough use case, may be worth having a helper like (untested):

```typescript
class SimpleValueQuery<T> extends Query<[], T> {
  constructor(name: string, public readonly initialValue: T) {
    super(name)
  }

  newValueRef(): ({ value: T }) {
    const v = { value: this.initialValue }
    // Somehow the calling workflow is known to this setter?
    this.handle = () => v.value
    return v
  }
}
```

Reply:
I guess we can have a `QueryVariable` or `QueryableState` class in the global scope.
The limitation is that the initial value needs to be initialized before workflow execution and can't be based on workflow type or arguments for example.

```ts
export isBlockedQuery = new QueryableState('isBlocked', true);

export async function myWorkflow1() {
  unblock.handler = (isBlockedQuery.value = false); // value is bound to this workflow
  await condition(() => !isBlockedQuery.value);
}
```

Reply:
That is similar to what I was thinking except I don't see where the query handler is installed (inside the `value` setter on first invocation)? In my head I had the below that was not reliant on behaviors in accessors:

```typescript
export isBlockedQuery = new SimpleValueQuery('isBlocked', true)

export async function myWorkflow1() {
  const blocked = isBlockedQuery.newValueRef()
  unblock.handler = () => { blocked.value = false }
  await condition(() => !blocked.value)
}
```

So `newValueRef` is what hooks up the query handler. Basically the same thing then, cool.